crypto/bn/bn_mod.c

```diff
@@ -110,6 +111,61 @@ int BN_mod_sub(BIGNUM *r, const BIGNUM *a, const BIGNUM *b, const BIGNUM *m,
     return BN_nnmod(r, r, m, ctx);
 }
 
+/*
+ * BN_mod_sub variant that may be used if both a and b are non-negative,
+ * a is less than m, while b is of same width as m. It's implemented as
+ * subtraction followed by two conditional additions.
+ */
+int bn_mod_sub_fixed_top(BIGNUM *r, const BIGNUM *a, const BIGNUM *b,
+                         const BIGNUM *m)
+{
+    size_t i, ai, bi, mtop = m->top;
+    BN_ULONG borrow, carry, ta, tb, mask, *rp;
+    const BN_ULONG *ap, *bp;
+
+    if (bn_wexpand(r, mtop) == NULL)
+        return 0;
+
+    rp = r->d;
+    ap = a->d != NULL ? a->d : rp;
+    bp = b->d != NULL ? b->d : rp;
+
```

Comment:
Mostly checking this out loud, though I wouldn't object if you put it in a comment:

a is reduced and b is the same bit width, so we know:
0 <= a < m
0 <= b < 2^w <= 2*m

Thus the first subtraction gives:
    2*m < r = a - b < m - 0

Thus two conditional additions does the job.